File: xiaoming/0917/insertionSortList.java

public class insertionSortList {
	
	public static void main(String[] args) {
		
		ListNode helper = new ListNode();
		ListNode l1 = new ListNode(20);
		helper.next = l1;
		l1.next = new ListNode(6);
		l1 = l1.next;	
		l1.next = new ListNode(8);
		l1 = l1.next;
		l1.next = new ListNode(2);
		l1 = l1.next;
		l1.next = new ListNode(9);
		l1 = l1.next;
		l1.next = new ListNode(15);
		l1 = l1.next;
		l1.next = new ListNode(60);
		
		ListNode l2 = insertionSortListt(helper.next);
		
		while (l2 != null) {
			System.out.println(l2.val+ "  \n");
			l2 = l2.next;	
		}
		
	}
		
	public static ListNode insertionSortListt(ListNode head) {
		if (head == null)  return null;
		ListNode helper = new ListNode();
		//helper.next = head;
		ListNode pre = helper;
		ListNode cur = head;
		while (cur != null) {
			
			ListNode next = cur.next;
			pre = helper;
			while(pre.next != null && pre.next.val <= cur.val) {
				pre = pre.next;
			}
			cur.next = pre.next;
			pre.next = cur;			
			cur = next;
		}
		return helper.next;
	}
}
	
	
	
	class ListNode {
  int val;
  ListNode next;
  ListNode() { 
  	val = 0;
  	next = null;
  }
  ListNode(int x) { 
  	val = x;
  	next = null;
  }
}

File: Sen/Insertion_Sort_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
        	return head

        helper = ListNode(0)
        cur = head
        pre = helper
        next = None

        while cur:
        	next = cur.next
        	while pre.next and pre.next.val < cur.val:
        		pre = pre.next
        	cur.next = pre.next
        	pre.next = cur
        	pre = helper 
        	cur = next 

        return helper.next 

    def insertionSortList(self, head):
    	newhead = ListNode(0)
    	newhead.next = head 
    	pre, cur = newhead, head
    	while cur:
    		if cur.next and cur.next.val < cur.val:
    			while pre.next and pre.next.val < cur.next.val:
    				pre = pre.next
    			tmp = pre.next
    			pre.next = cur.next
    			cur.next = cur.next.next
    			pre.next.next = tmp
    			pre = newhead
    		else:
    			cur = cur.next
    	return newhead.next



